076-moving-data-between-two-jbod-disks-using-cruise-control.md

```diff
@@ -0,0 +1,85 @@
+This endpoint will allow us to move the data between two JBOD disks. 
+
+## Current situation
+
+Currently, we get a multiple requests from community users to add the ability for moving all Kafka logs between two disks on the JBOD storage array. This feature can be useful in following scenarios:
+- The current disk is too small and the user wants to use a bigger one, or vice versa.
+- When we want to use a different Storage Class with different parameters or different storage types.
+- In case of disk removal to reduce the total storage.
+
+For now, we can do this using the Kafka CLI tool i.e. `kafka-reassign-partitions.sh` tool, but it takes a lot of manual steps which is time-consuming and not so user-friendly.
+
+## Motivation
+
+We should introduce the logic to Strimzi to leverage Cruise Control integration and make it possible to move the data between two JBOD disks.
+This feature will also allow us to remove the disks without the loss of data.
+
+## Proposal
+
+Cruise Control provides `remove_disks` HTTP REST endpoint to move replicas from a specified disk to other disks of the same broker. It always uses intra-broker re-balancing.
+This endpoint triggers a rebalancing operation by moving replicas in a size-based manner to the remaining disks, from the largest to the smallest, while checking the following constraint:
+```sh
+1 - (remainingUsageAfterRemoval / remainingCapacity) > errorMargin
+```
+where:
+```sh
+remainingUsageAfterRemoval = current usage for remaining disks + additional usage from removed disks
+remainingCapacity = sum of capacities of the remaining disks
+errorMargin = configurable property (default 0.1); it makes sure that a disk percentage is always free when moving replicas
+```
+
+In order to use the `remove_disks` endpoint in the Strimzi cluster operator, it would be added to the [`CruiseControlApi`](https://github.com/strimzi/strimzi-kafka-operator/blob/main/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/cruisecontrol/CruiseControlApi.java) interface and developing the corresponding implementation.
+
+### Implementation
+
+To implement this feature, we will be adding a new mode to the `KafkaRebalanceMode` class.
+* `remove-disks`: It moves replicas from a specified disk to other disks of the same broker. It always uses intra-broker re-balancing.
+You can use this mode by changing the `spec.mode` to `remove-disks` in the `KafkaRebalance` resource.
+
+A `KafkaRebalance` custom resource would look like this.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaRebalance
+metadata:
+  name: my-rebalance
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  mode: remove-disks
+  brokersAndVolumes:
+    - brokerId: 0
+      volumeIds: [1, 2]
+    - brokerId: 2
+      volumeIds: [1]
```

Comment:
Are all these combinations packed into a single Cruise Control request? Is there any limit for the number of items this list can have?

Reply:
Yes they will go all together in the query parameter, so something like `/remove_disks?brokerid_and_logdirs=0-/tmp/kafka-logs-1,0-/tmp/kafka-logs-2,2=-/tmp/kafka-logs-1`. From the CC code I didn't find any limit. It just parses a string using `,` as separator across brokers/volumes and `-` between broker id and log dir.

Reply:
I gues the URL size will have some limit?

Reply:
Good point. I didn't find anything useful in the official HTTP 1.1 RFC. I just found different length for different browsers which of course we don't use. From 2000 up to more and more.
Are you wondering if we should put a limit on how many brokers and volumes could be involved in order to limit that query parameter length?

Reply:
Anyway, I guess if it has a separate item for each brokerID-logDir pair it doesn't matter as we cannot change that. I was mostly wondering if it would make sense to have a list of brokers or range for the `brokerId`. But it does not seem so. Thanks for the explanation @ppatierno